libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -6122,3 +6118,131 @@ nubOrdMemberSpec message xs =
         , "The input is the empty list."
         ]
     )
+
+-- ==============================================================================
+-- Experimental changes to conformsToSpecM
+
+runTermE :: Env -> Term fn a -> Either (NE.NonEmpty String) a
+runTermE env = \case
+  Lit a -> Right a
+  V v -> case lookupEnv env v of
+    Just a -> Right a
+    Nothing -> Left (pure ("Couldn't find " ++ show v ++ " in " ++ show env))
+  App f ts -> do
+    vs <- mapMList (fmap Identity . runTermE env) ts
+    pure $ uncurryList_ runIdentity (sem f) vs
+
+checkPredsE ::
+  FunctionLike fn =>
+  NE.NonEmpty String ->
+  Env ->
+  [Pred fn] ->
+  Maybe (NE.NonEmpty String)
+checkPredsE msgs env ps =
+  case catMaybes (fmap (checkPredE env msgs) ps) of
+    [] -> Nothing
+    (x : xs) -> Just (NE.nub (foldl (<>) x xs))
```

Comment:
You should know better, `foldl` should never be used!
It is `foldl'` that is the function for folding. 

However, in this case not having `Monoid` is not an issue. `NonEmpty` has a `Semigroup` instance, so it can easily be switched to:
```suggestion
    (x : xs) -> Just (NE.nub (sconcat (x NE.:| xs))
```

Please apply the same suggestion to all such places. Last thing we want is a space leak to to the horrible `foldl` from `Prelude`

Reply:
done (4 places in all)